### PR TITLE
Spruce up docs for flake8-pyi rules

### PR DESCRIPTION
**Pull request by AlexWaygood**
This started off as an attempt to clarify some of the confusion in https://github.com/astral-sh/ruff/issues/9810#issuecomment-1999339083 and https://github.com/astral-sh/ruff/issues/9810#issuecomment-1999349880, but I figured I would take a more comprehensive look at the docs for the flake8-pyi rules while I was about it.

Summary of the changes:

- Improve clarity over the motivation for some rules
- Improve links to external references. In particular, reduce links to PEPs, as PEPs are generally historical documents rather than pieces of living documentation. Where possible, it's better to link to the [official typing spec](https://typing.readthedocs.io/en/latest/spec/), the other docs at https://typing.readthedocs.io/en/latest/, or the docs at https://docs.python.org/3/library/typing.html.
- Add some missing imports to code examples
- Use more concise language in a few places

### CODE REVIEW

**Comment by AlexWaygood** [comment]
I haven't reviewed the docs for all the flake8-pyi rules yet -- only up to `redundant_numeric_union.rs` -- but I'll cut scope here and tackle the others in a separate PR, so that this is easier to review.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/any_eq_ne_annotation.rs [hunk 1]
```diff
@@ -8,16 +8,20 @@ use crate::checkers::ast::Checker;
...
+/// code, you should generally only use `Any` when the semantics of your code
+/// would otherwise be inexpressible to the type checker.
 ///
-/// The semantics of `__eq__` and `__ne__` are such that the `obj` parameter
-/// should be any type, as opposed to a dynamically typed value. Therefore, the
-/// `object` type annotation is more appropriate.
+/// The expectation in Python is that a comparison of two arbitrary objects
+/// using `==` or `!=` should never raise an exception. As such, the methods
+/// implementing these comparison operators -- `__eq__` and `__ne__` -- should
+/// use `object` rather than `Any` for their second parameter.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/any_eq_ne_annotation.rs:24]: This is probably on me but I don't understand the relation between "`==` or` !=` should never raise an exception" and why the method should then use `object` rather than `Any`. How does using `object` guarantee against exceptions (other than that it might catch a few more typing errors)

File: crates/ruff_linter/src/rules/flake8_pyi/rules/collections_named_tuple.rs [hunk 2]
```diff
@@ -13,16 +13,17 @@ use crate::checkers::ast::Checker;
...
-/// The class generated by subclassing `typing.NamedTuple` is equivalent to
-/// `collections.namedtuple`, with the exception that `typing.NamedTuple`
-/// includes an `__annotations__` attribute, which allows type checkers to
-/// infer the types of the fields.
+/// Inheriting from `typing.NamedTuple` creates a custom `tuple` subclass in
+/// the same way as using the `collections.namedtuple` factory function.
+/// However, using `typing.NamedTuple` allows you to provide a type annotation
+/// for each field in the class. This means that type checkers will have much
+/// more information to work with, and will be able to analyze your code far
+/// more precisely.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/collections_named_tuple.rs:21]:
```suggestion
/// However, using `typing.NamedTuple` allows you to provide a type annotation
/// for each field in the class. This means that type checkers will have
/// more information to work with, and will be able to analyze your code
/// more precisely.
```

Maybe we can rephrase it to say that `namedtuple` types all fields as `Any` (I guess?) and that `typing.NamedTuple` allows you to specify the field types instead.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/complex_assignment_in_stub.rs [hunk 3]
```diff
@@ -20,18 +20,28 @@ use crate::checkers::ast::Checker;
...
+/// from typing import TypeAlias
+///
 /// a: TypeAlias = int
 /// b: TypeAlias = int
 ///
 ///
-/// class a:
-///     b: int
+/// class Klass:
+///     X: TypeAlias = int
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/complex_assignment_in_stub.rs:44]: Noob question. Is this preferred over `type X = int` or is that syntax not allowed in classes/stubs?

File: crates/ruff_linter/src/rules/flake8_pyi/rules/future_annotations_in_stub.rs [hunk 4]
```diff
@@ -10,9 +10,10 @@ use crate::checkers::ast::Checker;
 /// statement in stub files.
 ///
 /// ## Why is this bad?
-/// Stub files are already evaluated under `annotations` semantics. As such,
-/// the `from __future__ import annotations` import statement has no effect
-/// and should be omitted.
+/// Stub files natively support forward references in all contexts, as stubs are
+/// never executed at runtime. (They should be thought of as "data files" for
+/// type checkers.) As such, the `from __future__ import annotations` import
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/future_annotations_in_stub.rs:15]: This might be correct but it's weird that the `.` is inside of the parentheses :sweat_smile: 

```suggestion
/// type checkers). As such, the `from __future__ import annotations` import
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs [hunk 5]
```diff
@@ -15,24 +15,46 @@ use crate::checkers::ast::Checker;
 /// `__iter__` methods should always should return an `Iterator` of some kind,
 /// not an `Iterable`.
 ///
-/// In Python, an `Iterator` is an object that has a `__next__` method, which
-/// provides a consistent interface for sequentially processing elements from
-/// a sequence or other iterable object. Meanwhile, an `Iterable` is an object
-/// with an `__iter__` method, which itself returns an `Iterator`.
+/// In Python, an `Iterable` is an object that has an `__iter__` method; an
+/// `Iterator` is an object that has `__iter__` and `__next__` methods. All
+/// `__iter__` methods are expected to return `Iterator`s in Python. Type
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs:20]: I would probably remove the second "in Python", assuming that's clear from a python linter.
```suggestion
/// `__iter__` methods are expected to return `Iterator`s. Type
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs [hunk 6]
```diff
@@ -15,24 +15,46 @@ use crate::checkers::ast::Checker;
...
+/// >>> x = [42]
+/// >>> isinstance(x, collections.abc.Iterable)
+/// True
+/// >>> isinstance(x, collections.abc.Iterator)
+/// False
+/// >>> next(x)
+/// Traceback (most recent call last):
+///  File "<stdin>", line 1, in <module>
+/// TypeError: 'list' object is not an iterator
+/// >>> y = iter(x)
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs:39]: What I understood from the description "by calling `iter` on the list" is that you write `y = x.iter()`. 

Maybe that's because of my German background but "on" sounds like it's a method on list. It might already be sufficient to change the example to `iter(...)` or `iter(x)` or use another word than `on`.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs [hunk 7]
```diff
@@ -15,24 +15,46 @@ use crate::checkers::ast::Checker;
...
+/// True
+/// >>> isinstance(y, collections.abc.Iterator)
+/// True
+/// >>> next(y)
+/// 42
+/// ```
+///
+/// Using `Iterable` rather than `Iterator` as a return type for an `__iter__`
+/// methods would imply that you would not necessarily be able to call `next()`
+/// on the returned object, violating the expectations of the interface.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs:50]: Nit: I'm okay with either.

```suggestion
/// on the returned object, violating the contract of the interface.
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/no_return_argument_annotation.rs [hunk 8]
```diff
@@ -9,27 +9,25 @@ use crate::checkers::ast::Checker;
...
+/// Checks for uses of `typing.NoReturn` (and `typing_extensions.NoReturn`) for
+/// parameter annotations.
 ///
 /// ## Why is this bad?
-/// Prefer `typing.Never` (or `typing_extensions.Never`) over `typing.NoReturn`,
-/// as the former is more explicit about the intent of the annotation. This is
-/// a purely stylistic choice, as the two are semantically equivalent.
+/// Prefer `typing.Never` (or `typing_extensions.Never`) over `typing.NoReturn`
+/// for parameter annotations. This is a purely stylistic choice, as the two
+/// are semantically equivalent.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/no_return_argument_annotation.rs:18]: I think this misses an explanation of why it's more idiomatic (or a preferred stylistic choice) to use  `Never` over `NoReturn`. I assume it's mainly that `NoReturn` is intended for function return types and everything else should use `Never`?

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_literal_union.rs [hunk 9]
```diff
@@ -13,30 +13,28 @@ use crate::checkers::ast::Checker;
...
-/// A: Literal["A"] | str
+/// x: Literal["A", b"B"] | str
 /// ```
 ///
 /// Use instead:
 /// ```python
 /// from typing import Literal
 ///
-/// A: Literal["A"]
+/// x: Literal["A"] | str
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_literal_union.rs:37]: Shouldn't this be `str`?

> For example, `Literal["A"] | str` is equivalent to `str`, and

**Review by MichaReiser — APPROVED** [review]
Nice improvements. I learned a couple of things by reading through your new documentation.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs (hunk 7 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs:50]: I weakly prefer "expectations of the interface", so I'll stick with my current wording here 👍

File: crates/ruff_linter/src/rules/flake8_pyi/rules/any_eq_ne_annotation.rs (hunk 1 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/any_eq_ne_annotation.rs:24]: Basically, for any two objects `x` and `y`, `x == y` should _never_ raise an exception -- calling `x.__eq__(y)` should be safe regardless of what the types are. That means that `object` is the better type here, as it's Python's true "top type" from a type theory perspective.

`Any` should be reserved for situations where you either want to allow unsafe behaviour for some reason, or for situations where your code is safe, but the true type that a function accepts is inexpressible. That should never be the case for `__eq__` or `__ne__` methods -- `object` should do just fine in all cases.

Now, how can I put that more succinctly 😅

File: crates/ruff_linter/src/rules/flake8_pyi/rules/complex_assignment_in_stub.rs (hunk 3 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/complex_assignment_in_stub.rs:44]: `type X = int` is allowed in classes/stubs, but it's new syntax that's only available on Python 3.12+ -- I think it's better to use syntax here that all users of Python will be able to use

File: crates/ruff_linter/src/rules/flake8_pyi/rules/future_annotations_in_stub.rs (hunk 4 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/future_annotations_in_stub.rs:15]: The rule is that if it's a complete sentence inside the parentheses, the period goes inside the parentheses, but if it's only part of a sentence inside the parentheses, the period goes outside: https://www.masterclass.com/articles/period-inside-or-outside-parentheses

File: crates/ruff_linter/src/rules/flake8_pyi/rules/no_return_argument_annotation.rs (hunk 8 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/no_return_argument_annotation.rs:18]: It's basically just because it's really confusing to see a type called `NoReturn` in a parameter annotation -- that's the only reason we added the `Never` type in Python 3.11 😄 It has exactly the same semantics as `Noreturn`.

I'll try to flesh this out a little more...

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_literal_union.rs (hunk 9 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_literal_union.rs:37]: Whoops, good catch. This should be `Literal[b"B"] | str`.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/collections_named_tuple.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/collections_named_tuple.rs]: > Maybe we can rephrase it to say that `namedtuple` types all fields as `Any` (I guess?)

That's correct, but I wonder if we really need to go into that much detail here... I think simply saying that it will make type-checking more accurate is okay for our purposes here, and helps avoid too much information overload

File: crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs (hunk 6 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs:39]: You're right that this was unclear -- I changed the wording from

```
calling `iter()` on the list
```

to

```
passing the list to `iter()`
````
